### PR TITLE
V4100

### PR DESCRIPTION
**Pull request by Obaied**
- Add event tracking success/failure delegates
- Add session tracking success/failure delegates
- Add deeplink listener
- Support for sessions parameters (callback & partners)
- Update submodules (Android & iOS) to 4.10.2 & 4.10.0 respectively
- Add BASH build scripts for both Android & iOS
- Update Android submodule build script to use BASH and build with Gradle
- Update iOS submodule build script to use BASH

### CODE REVIEW

File: src/iOS/AdjustCordova.m [hunk 1]
```diff
@@ -38,7 +54,55 @@ - (void)adjustAttributionChanged:(ADJAttribution *)attribution {
...
+    NSDictionary *dict = [sessionFailureResponseData dictionary];
+
+    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:dict];
+    pluginResult.keepCallback = [NSNumber numberWithBool:YES];
+
+    [self.commandDelegate sendPluginResult:pluginResult callbackId:sessionFailureCallbackId];
+}
+
+- (BOOL)adjustDeeplinkResponse:(NSURL *)deeplink {
+    //NSMutableDictionary *dict = [NSMutableDictionary new];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: commented code should be removed

File: src/iOS/AdjustCordova.m [hunk 2]
```diff
@@ -49,12 +113,18 @@ - (void)create:(CDVInvokedUrlCommand *)command {
...
+
+    NSString *userAgent = [[command.arguments objectAtIndex:0] objectForKey:KEY_USER_AGENT];
+    double delayStart = [[[command.arguments objectAtIndex:0] objectForKey:KEY_DELAY_START] doubleValue];
+
     ADJConfig *adjustConfig = [ADJConfig configWithAppToken:appToken environment:environment];
 
     if ([adjustConfig isValid]) {
         // Log level
         if ([self isFieldValid:logLevel]) {
-            [adjustConfig setLogLevel:[ADJLogger LogLevelFromString:[logLevel lowercaseString]]];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 3]
```diff
@@ -93,11 +202,11 @@ - (void)trackEvent:(CDVInvokedUrlCommand *)command {
     NSMutableArray *partnerParameters = [[NSMutableArray alloc] init];
 
     for (id item in [[command.arguments objectAtIndex:0] objectForKey:KEY_CALLBACK_PARAMETERS]) {
-        [callbackParameters addObject:item];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 4]
```diff
@@ -93,11 +202,11 @@ - (void)trackEvent:(CDVInvokedUrlCommand *)command {
...
+                        [callbackParameters addObject:item];
     }
 
     for (id item in [[command.arguments objectAtIndex:0] objectForKey:KEY_PARTNER_PARAMETERS]) {
-        [partnerParameters addObject:item];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 5]
```diff
@@ -130,16 +239,16 @@ - (void)trackEvent:(CDVInvokedUrlCommand *)command {
                 [adjustEvent setReceipt:[receipt dataUsingEncoding:NSUTF8StringEncoding] transactionId:transactionId];
             } else {
                 if ([self isFieldValid:transactionId]) {
-                    [adjustEvent setTransactionId:transactionId];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 6]
```diff
@@ -130,16 +239,16 @@ - (void)trackEvent:(CDVInvokedUrlCommand *)command {
...
+         [adjustEvent setTransactionId:transactionId];
 
-                    isTransactionIdSet = YES;
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 7]
```diff
@@ -130,16 +239,16 @@ - (void)trackEvent:(CDVInvokedUrlCommand *)command {
...
+         isTransactionIdSet = YES;
                 }
             }
         }
 
         if (NO == isTransactionIdSet) {
             if ([self isFieldValid:transactionId]) {
-                [adjustEvent setTransactionId:transactionId];
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: this change removed correct number of spaces for indentation

File: src/iOS/AdjustCordova.m [hunk 8]
```diff
@@ -204,11 +313,85 @@ - (void)onResume:(CDVInvokedUrlCommand *)command {
...
+- (void)setDeeplinkCallback:(CDVInvokedUrlCommand *)command {
+    deeplinkCallbackId = command.callbackId;
+}
+
+- (void)addSessionCallbackParameter:(CDVInvokedUrlCommand *)command {
+    NSString *key = [command argumentAtIndex:0 withDefault:nil];
+    NSString *value = [command argumentAtIndex:1 withDefault:nil];
+
+    if (!([self isFieldValid:key]) || !([self isFieldValid:value]) ) {
+    return;
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: should indented inside the block

File: src/iOS/AdjustCordova.m [hunk 9]
```diff
@@ -204,11 +313,85 @@ - (void)onResume:(CDVInvokedUrlCommand *)command {
...
+    }
+
+    [Adjust addSessionCallbackParameter:key value:value];
+}
+
+- (void)removeSessionCallbackParameter:(CDVInvokedUrlCommand *)command {
+    NSString *key = [command argumentAtIndex:0 withDefault:nil];
+
+    if (!([self isFieldValid:key])) {
+    return;
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: should indented inside the block

File: src/iOS/AdjustCordova.m [hunk 10]
```diff
@@ -204,11 +313,85 @@ - (void)onResume:(CDVInvokedUrlCommand *)command {
...
+- (void)resetSessionCallbackParameters:(CDVInvokedUrlCommand *)command {
+    [Adjust resetSessionCallbackParameters];
+}
+
+- (void)addSessionPartnerParameter:(CDVInvokedUrlCommand *)command {
+    NSString *key = [command argumentAtIndex:0 withDefault:nil];
+    NSString *value = [command argumentAtIndex:1 withDefault:nil];
+
+    if (!([self isFieldValid:key]) || !([self isFieldValid:value]) ) {
+    return;
```
Comment (nonelse) [inline src/iOS/AdjustCordova.m]: should indented inside the block

File: www/adjust_config.js [hunk 11]
```diff
@@ -1,34 +1,94 @@
 function AdjustConfig(appToken, environment) {
     // iOS & Android
+    console.log("appToken: " + appToken);
```
Comment (nonelse) [inline www/adjust_config.js]: there should not be debug message that cannot be disable in production

**Comment by uerceg** [comment]
@adjust/integration

**Pull request merged**